pkg/assembler/backends/neo4j/pkg.go

```diff
@@ -339,63 +315,61 @@ func (c *neo4jClient) Packages(ctx context.Context, pkgSpec *model.PkgSpec) ([]*
 			var firstMatch bool = true
 			queryValues := map[string]any{}
 
-			if pkgSpec.MatchOnlyEmptyQualifiers != nil && !*pkgSpec.MatchOnlyEmptyQualifiers {
-				sb.WriteString("MATCH (n:Pkg)-[:PkgHasType]->(type:PkgType)-[:PkgHasNamespace]->(namespace:PkgNamespace)-[:PkgHasName]->(name:PkgName)-[:PkgHasVersion]->(version:PkgVersion)-[:PkgHasQualifier]->(qualifier:PkgQualifier)")
-			} else {
-				sb.WriteString("MATCH (n:Pkg)-[:PkgHasType]->(type:PkgType)-[:PkgHasNamespace]->(namespace:PkgNamespace)-[:PkgHasName]->(name:PkgName)-[:PkgHasVersion]->(version:PkgVersion)")
-			}
-			if pkgSpec.Type != nil {
+			sb.WriteString("MATCH (n:Pkg)-[:PkgHasType]->(type:PkgType)-[:PkgHasNamespace]->(namespace:PkgNamespace)-[:PkgHasName]->(name:PkgName)-[:PkgHasVersion]->(version:PkgVersion)")
 
+			if pkgSpec.Type != nil {
 				matchProperties(&sb, firstMatch, "type", "type", "$pkgType")
 				firstMatch = false
 				queryValues["pkgType"] = pkgSpec.Type
 			}
-			if pkgSpec.Namespace != nil {
 
+			if pkgSpec.Namespace != nil {
 				matchProperties(&sb, firstMatch, "namespace", "namespace", "$pkgNamespace")
 				firstMatch = false
 				queryValues["pkgNamespace"] = pkgSpec.Namespace
 			}
-			if pkgSpec.Name != nil {
 
+			if pkgSpec.Name != nil {
 				matchProperties(&sb, firstMatch, "name", "name", "$pkgName")
 				firstMatch = false
 				queryValues["pkgName"] = pkgSpec.Name
 			}
-			if pkgSpec.Version != nil {
 
+			if pkgSpec.Version != nil {
 				matchProperties(&sb, firstMatch, "version", "version", "$pkgVerion")
 				firstMatch = false
 				queryValues["pkgVerion"] = pkgSpec.Version
 			}
 
 			if pkgSpec.Subpath != nil {
-
 				matchProperties(&sb, firstMatch, "version", "subpath", "$pkgSubpath")
 				firstMatch = false
 				queryValues["pkgSubpath"] = pkgSpec.Subpath
 			}
 
 			if pkgSpec.MatchOnlyEmptyQualifiers != nil && !*pkgSpec.MatchOnlyEmptyQualifiers {
-
 				if len(pkgSpec.Qualifiers) > 0 {
-					for _, qualifier := range pkgSpec.Qualifiers {
-						qualifierKey := removeInvalidCharFromProperty(qualifier.Key)
-						matchProperties(&sb, firstMatch, "qualifier", qualifierKey, "$"+qualifierKey)
-						firstMatch = false
-						queryValues[qualifierKey] = qualifier.Value
+					qualifiersMap := map[string]string{}
+					keys := []string{}
+					for _, kv := range pkgSpec.Qualifiers {
+						key := removeInvalidCharFromProperty(kv.Key)
+						qualifiersMap[key] = *kv.Value
+						keys = append(keys, key)
+					}
+					sort.Strings(keys)
+					qualifiers := []string{}
+					for _, k := range keys {
+						qualifiers = append(qualifiers, k, qualifiersMap[k])
 					}
+					matchProperties(&sb, firstMatch, "version", "qualifier_list", "$qualifier")
+					firstMatch = false
+					queryValues["qualifier"] = qualifiers
 				}
 			} else {
-				matchNotEdge(&sb, firstMatch, "version", "PkgHasQualifier", "PkgQualifier")
+				matchProperties(&sb, firstMatch, "version", "qualifier_list", "[]")
```

Comment:
should this be an empty `[]string{}` or does it not matter?

Reply:
```
matchProperties(sb, firstMatch, "depVersion", "qualifier_list", "$depQualifierList")
firstMatch = false
queryValues["depQualifierList"] = []string{}
```

Reply:
Shouldn't matter, but updated for consistency